File: January2022/Python/CarPooling.py

#Author :  Karthik K (https://www.linkedin.com/in/karthikmunirathinam/)

class Solution:
    #TC : O(n)
    #SC : O(1) 
    
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        occupancy=[0]*1001
        for index in range(len(trips)):
            occupancy[trips[index][1]]+=trips[index][0]
            occupancy[trips[index][2]]-=trips[index][0]
        currCapacity=0
        for currOccupancy in occupancy:
            currCapacity+=currOccupancy
            if currCapacity>capacity:
                return False
        return True
        

File: January2022/Java/CarPooling.java
class Solution {
	// TC : O(n)
	// SC : O(1)
	public boolean carPooling(int[][] trips, int capacity) {

		int[] occupancy = new int[1001];
		for (int i = 0; i < trips.length; i++) {
			occupancy[trips[i][1]] += trips[i][0];
			occupancy[trips[i][2]] -= trips[i][0];
		}
		int currCapacity = 0;
		for (int currOccupancy : occupancy) {
			currCapacity +=   currOccupancy;
			if (currCapacity>capacity) {
				return false;
			}
		}

		return true;
	}
}